gsoc-2025/template-hooks.md

```diff
@@ -0,0 +1,122 @@
+---
+parent: DLang GSoC 2025 Project Ideas
+nav_order: 1
+---
+
+# Translate DRuntime Hooks to Templates
+
+**Mentor:** [Teodor Duțu](teodor.dutu@gmail.com)
+
+| Spec | Details |
+|-|-|
+| **Difficulty Level** | hard |
+| **Project Duration** | 350 hours |
+| **Number of Contributors** | 2 |
+| **Prerequisites** | Familiarity with compilers, build systems; low-level programming |
+
+## Description
+
+High-level language constructs are often compiled into lower-level function calls that implement the same functionality.
+This process is called _lowering_
+It compiler design by offloading complex code into simpler function calls.
+These functions are implemented in D's runtime library (called DRuntime) and are commonly called [_runtime hooks_](https://wiki.dlang.org/Runtime_Hooks).
+
+Below is an example of such a runtime hook:
+
+```d
+struct S { ... }
+
+S[3] a, b;
+
+// Original code
+b = a;  // b is a copy of a 
+
+// Resulting DRuntime hook
+_d_arrayassign_l(b, a)  // copies a into b
+```
+
+`_d_arrayassign_l` handles array (re)allocation and assignment operators according to the types of `a` and `b` (`S[3]`) to simplify the compiler's work.
+
+As shown in the example above, runtime hooks require type information, such as which assignment operator to call and the arrays' sizes.
+D supports templates, but runtime hooks pre-date the introduction of templates to D.
+Therefore, they retrieve the necessary type information at runtime by receiving an extra argument, whose type is [`TypeInfo`](https://dlang.org/library/object/type_info.html).
+This object contains the size, constructors, destructors and overloaded operators of a given type.
+However, because this information is processed at run time, this approach is slower to the alternative of having the compiler send it to the runtime hook via template arguments.
+Due to D's high flexibility regarding metaprogramming, translating each hook to a template function would allow its code to specialise TODO
+
+So far the following hooks have been converted to templates:
+
+```text
+_d_arrayctor
+_d_arraysetctor
+_d_arrayassign
+_d_arrayassign_l
+_d_arrayassign_r
+_d_delstruct
+_d_newThrowable
+_d_arrayappendT
+_d_arrayappendcTX
+_d_arraycatT
+_d_arraycatnTX
+_d_newitemiT
+_d_newitemT
+_d_newitemU
+_d_newclass
+_d_newarrayiT
+_d_newarrayT
+_d_newarrayU
+_d_newarraymTX
+_d_newarrayOpT
+```
```

Comment:
Having the full list seems unnecessary to me. It gives the impression that the project is already done.